### ISSUE TITLE
ExecuteV2() produces the same output after the first call 

### ISSUE THREAD

**Issue by borarak** [NVIDIA/TensorRT#1372]
## Description

I'm trying to inference a few images with a  batch size = 1 by calling executeV2() on them one by one. The pseudo code is as follows:

for images in my image_array:
	1. read image into host memory
	2. allocate input and output buffers on device
	3. Copy image data from host to deivce
	4. executeV2()
	5. Copy data from device to host
	6. Deallocate device buffers

The actual code is below: 

```
int b_size = 1;
float *emb = new float[b_size * 2048];
float  *data = new float[b_size * 3 * INPUT_H_MNET * INPUT_W_MNET];

//inference loop
for(const std::pair<int, cv::Mat> &b_pair: p_images){
  int b = b_pair.first;
  cv::Mat resized = b_pair.second;
  int i = 0;
  
 //read images from cv::Mat 
 for (int row = 0; row < INPUT_H_MNET; ++row) {
	  uchar* uc_pixel = resized.data + row * resized.step;
	  for (int col = 0; col < INPUT_W_MNET; ++col) {
		  data[i] = (float)uc_pixel[2] / 255.0;
		  data[i + INPUT_H_MNET * INPUT_W_MNET] = (float)uc_pixel[1] / 255.0;
		  data[i + 2 * INPUT_H_MNET * INPUT_W_MNET] = (float)uc_pixel[0] / 255.0;
		  uc_pixel += 3;
		  ++i;
	  }
  }
  
  //allocate
  CUDA_CHECK(cudaMalloc(&this->trackBuffers[this->inputIndex], b_size * 3 * INPUT_H_MNET * INPUT_W_MNET * sizeof(float)));
  CUDA_CHECK(cudaMalloc(&this->trackBuffers[this->outputIndex], b_size * OUTPUT_MNET * sizeof(float)));
  CUDA_CHECK(cudaMemcpyAsync(this->trackBuffers[this->inputIndex], data, b_size * 3 * INPUT_H_MNET * INPUT_W_MNET * sizeof(float), cudaMemcpyHostToDevice, this->streamTracker));	
  
  std::vector<void*> data_all = {this->trackBuffers[this->inputIndex], this->trackBuffers[this->outputIndex]};
  int res = this->contextTracker->executeV2(data_all.data()); // <<<<---- called here
  
  if(cudaMemcpyAsync(emb, this->trackBuffers[this->outputIndex], b_size * OUTPUT_MNET * sizeof(float), cudaMemcpyDeviceToHost, this->streamTracker) == cudaSuccess){
	  std::cout << "Copy back success!" << std::endl;
  }else{
	  std::cout << "Copy back failure!!!" << std::endl;
	  assert(0);
  }
  std::cout << "Results (AFTER TRT CALL)  : " << emb[0] << "," << emb[10] << "," << emb[20] << "," << emb[30] << "," ; //  <<<----- Results are the same result after the first iteration of the loop
  cudaStreamSynchronize(this->streamTracker);
  cudaFree(this->trackBuffers[this->inputIndex]);
  cudaFree(this->trackBuffers[this->outputIndex]);
}
```

After the first iteration of the loop, the `executeV2()` call always produces the same results identical to the result of the first call. I checked the input data fed to the buffers is different each time whereas the expected behavior will be to have different results each time i call executeV2(). Can somebody help me understand why I'm seeing the unexpected behavior? Thanks in advance!

## Environment

**TensorRT Version**: 7.2.1
**NVIDIA GPU**: RTX 2080 Super
**NVIDIA Driver Version**: 455.23.04
**CUDA Version**: 11.1
**CUDNN Version**: 8.0.5
**Operating System**: Ubuntu 18.04 LTS
**Python Version (if applicable)**: 
**Tensorflow Version (if applicable)**: 
**PyTorch Version (if applicable)**: 
**Baremetal or Container (if so, version)**:

**Comment by pranavm-nvidia** [comment]
I think you'd probably want to either change the `cudaMemcpy`s to be synchronous, or change the inference call (i.e. `executeV2`) to be asynchronous (`enqueueV2`). 

I'm not sure what the behavior is when you follow an asynchronous memcpy with `executeV2`, which has no knowledge of the stream being used to copy the inputs.

**Comment by borarak** [comment]
@pranavm-nvidia I should have mentioned I initially tried with `enqueueV2` with the following call replacing the `executeV2()` call
`int res = this->contextTracker->enqueueV2(this->trackBuffers, this->streamTracker, nullptr);`

This had similar results (i.e after the first iteration, the output buffer has always the same values). I double checked the input buffer is always different.

**Comment by pranavm-nvidia** [comment]
Are the outputs of the first iteration correct? Also, could you try moving the allocations out of the inference loop? That'll eliminate more variables.

**Comment by borarak** [comment]
I just checked that the outputs of the first iteration is incorrect too. It seems to produce the same values regardless of the input. I'm parsing from an ONNX file and I have updated the issue with the code that generates the engine.

**Comment by pranavm-nvidia** [comment]
Sounds like either the inputs aren't being properly copied to the device, or the outputs aren't being copied back. Do you see similar behavior with `trtexec --dumpOutput`?

**Comment by borarak** [comment]
@pranavm-nvidia I tried with `trtexec` and I have quite strange behavior.

1. `trtexec -onnx="my_onnx_file.onnx" --dumpOutput` -> always results in exactly the same output no matter how many time I run

Next, I create two different dummy inputs of the same size as my input (1x3x296x128) and try the following:

2. `trtexec -onnx="my_onnx_file.onnx" --loadInputs="input":"sample_input1.bin" --dumpOutput`
3. `trtexec -onnx="my_onnx_file.onnx" --loadInputs="input":"sample_input2.bin" --dumpOutput`

In both cases i get the same output but the result is different from (1)

**Comment by pranavm-nvidia** [comment]
That is really strange. Would you be able to share the ONNX model?

**Comment by borarak** [comment]
@pranavm-nvidia Yes, here are the files:
1. ONNX model file: https://drive.google.com/file/d/1Lndl7a1m2lAuVzQrY5Bew8JzZOZdlAtm/view?usp=sharing
2. Original PyTorch (MobileNet) file: https://drive.google.com/file/d/1Lndl7a1m2lAuVzQrY5Bew8JzZOZdlAtm/view?usp=sharing

**Comment by pranavm-nvidia** [comment]
Did a little debugging with [Polygraphy](https://github.com/NVIDIA/TensorRT/tree/master/tools/Polygraphy) and it seems to be an issue with the model (e.g. outputs are the same even when using ONNX-Runtime).

I dumped layer-wise outputs from two runs using different inputs (different `--seed`):
```bash
polygraphy run mnet_siam_op11.onnx --onnxrt --onnx-outputs mark all \
    --seed 0 --save-outputs outputs0.json 
polygraphy run mnet_siam_op11.onnx --onnxrt --onnx-outputs mark all \
    --seed 1 --save-outputs outputs1.json 
```

Then, comparing the layer-wise outputs:
```bash
polygraphy run --load-outputs outputs0.json outputs1.json 
```
we can see that the first layer where the outputs being to match is `324`:
```bash
[I]     Comparing Output: '547' (dtype=float32, shape=(1, 32, 128, 64)) with '547' (dtype=float32, shape=(1, 32, 128, 64)) | Tolerance: [abs=1e-05, rel=1e-05] | Checking elemwise error
[E]         FAILED | Difference exceeds tolerance (rel=1e-05, abs=1e-05)
[I]     Comparing Output: '324' (dtype=float32, shape=(1, 32, 128, 64)) with '324' (dtype=float32, shape=(1, 32, 128, 64)) | Tolerance: [abs=1e-05, rel=1e-05] | Checking elemwise error
[I]         onnxrt-runner-N0-07/15/21-10:04:03: 324 | Stats: mean=2.2567, std-dev=2.9025, var=8.4246, median=0, min=0 at (0, 0, 0, 0), max=6 at (0, 2, 0, 0)
[I]         onnxrt-runner-N0-07/15/21-10:03:54: 324 | Stats: mean=2.2567, std-dev=2.9025, var=8.4246, median=0, min=0 at (0, 0, 0, 0), max=6 at (0, 2, 0, 0)
[I]         Error Metrics: 324
[I]             Minimum Required Tolerance: elemwise error | [abs=0] OR [rel=0]
[I]             Absolute Difference | Stats: mean=0, std-dev=0, var=0, median=0, min=0 at (0, 0, 0, 0), max=0 at (0, 0, 0, 0)
[I]             Relative Difference | Stats: mean=nan, std-dev=nan, var=nan, median=nan, min=nan at (0, 0, 0, 0), max=nan at (0, 0, 0, 0)
[I]         PASSED | Difference is within tolerance (rel=1e-05, abs=1e-05)
```

Looking at the model in Netron, `324` is the output of a ReLU6. The immediately preceding convolution layer has some very large weights:

![large_weights](https://user-images.githubusercontent.com/49246958/125829572-73c5c116-70c6-44b5-abf9-08849326e740.png)

So most likely the outputs of this Conv are going outside the range `[0, 6]` and hence are clamped.

**Comment by borarak** [comment]
Thanks @pranavm-nvidia , some great insights there. I will debug more and find out why that's happening.

**Comment by borarak** [comment]
In the end, this was related to the C++ input not getting normalized in the same way as in Python. Closing and thanks for the help!
